src/main/scala/org/apache/spark/mllib/optimization/tfocs/DVectorFunctions.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.optimization.tfocs
+
+import org.apache.spark.mllib.linalg.{ BLAS, DenseVector, Vector }
+import org.apache.spark.mllib.optimization.tfocs.VectorSpace._
+
+/**
+ * Extra functions available on DVectors through an implicit conversion. DVectors are represented
+ * using RDD[Vector], and these helper functions apply operations to the values within each Vector
+ * of the RDD.
+ */
+private[tfocs] class DVectorFunctions(self: DVector) {
+
+  def mapElements(f: Double => Double): DVector = self.map(x => new DenseVector(x.toArray.map(f)))
+
+  def zipElements(other: DVector, f: (Double, Double) => Double): DVector =
+    self.zip(other).map({ x =>
+      if (x._1.size != x._2.size) {
+        throw new IllegalArgumentException("Can only zipElements DVectors with the same number " +
+          "of elements")
+      }
+      new DenseVector(x._1.toArray.zip(x._2.toArray).map(y => f(y._1, y._2))): Vector
```

Comment:
I believe I was previously zipping iterators created using x.toArray.toIterator, but I think I was asked to avoid using these iterators. I can convert to a while loop.